### PR TITLE
Check version

### LINKED ISSUE #213: Add release number in CLI, tell you when to update.
It would be neat if the release number was added to the CLI and a way to check for the latest release and tell you if you should update.

### PR DESCRIPTION
**Pull request by nol166**
### Describe in detail the problem you had and how this PR fixes it
Checks to see if there is a newer release available when running the CLI

### Is there an open issue you can link to?
Fixes #213.

### CODE REVIEW

File: packages/server/src/updater.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+export const getRecentRelease = (): Promise<any> => {
+	return new Promise<string>((resolve, rej) => {
+		const options = {
+			host: "api.github.com",
+			path: "/repos/codercom/code-server/releases",
+			method: "GET",
+			agent: false,
+			headers: {
+				"User-Agent": "test",
```
Comment (NGTmeaty) [inline packages/server/src/updater.ts]: We'll want to follow [Github's user-agent requirements](https://developer.github.com/v3/#user-agent-required) and change it to code-server.

File: packages/server/src/updater.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+			return -1;
+		}
+	}
+};
+
+export const getRecentRelease = (): Promise<any> => {
+	return new Promise<string>((resolve, rej) => {
+		const options = {
+			host: "api.github.com",
+			path: "/repos/codercom/code-server/releases",
```
Comment (NGTmeaty) [inline packages/server/src/updater.ts]: Looking at the API again, we could possibly [call /latest](https://developer.github.com/v3/repos/releases/#get-the-latest-release) and get just compare those names?

**Comment by NGTmeaty** [comment]
The Docker version we compile doesn't get a version number. Might be a different issue, but it'd be nice to get version checking in there too :)

**Comment by nol166** [comment]
> The Docker version we compile doesn't get a version number. Might be a different issue, but it'd be nice to get version checking in there too :)
> 

The docker image eventually inherits the source code from the binary so as soon as the CI catches up (if and when) this is merged it should work for the docker image as well.

File: packages/server/src/updater.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+			method: "GET",
+			agent: false,
+			headers: {
+				"User-Agent": "code-server",
+			},
+		};
+
+		https.get(options, function (res) {
+			if (res.statusCode !== 200) {
+				throw new Error("Failed to acquire release information");
```
Comment (lsmoura) [inline packages/server/src/updater.ts]: You're inside a Promise. Maybe you should let the caller knows by rejecting it?

```suggestion
				rej(new Error("Failed to acquire release information"));
```

File: packages/server/src/cli.ts [hunk 4]
```diff
@@ -139,6 +140,22 @@ if (isCli) {
 		}
 	}
 
+	let newestVersion = "";
+	if (process.env.VERSION) {
+		let currentVersion = process.env.VERSION;
+		const recentRelease = await getRecentRelease();
+
+		if (compareVersions(currentVersion, recentRelease)! <= 1) {
```
Comment (kylecarbs) [inline packages/server/src/cli.ts]: No need to compare versions. If the latest version !== the current version then an update is available.

File: packages/server/src/cli.ts [hunk 5]
```diff
@@ -139,6 +140,22 @@ if (isCli) {
 		}
 	}
 
+	let newestVersion = "";
```
Comment (kylecarbs) [inline packages/server/src/cli.ts]: Would be cleaner to do something like:
```
if (isCli) {
    const currentVersion = process.env.VERSION;
    if (!currentVersion) {
        return;
    }
    const recentRelease = await getRecentRelease();
    ...
}
```

File: packages/server/src/cli.ts [hunk 6]
```diff
@@ -139,6 +140,22 @@ if (isCli) {
...
+		const recentRelease = await getRecentRelease();
+
+		if (compareVersions(currentVersion, recentRelease)! <= 1) {
+			newestVersion = recentRelease;
+		} else {
+			newestVersion = currentVersion;
+		}
+	}
+	if (newestVersion === process.env.VERSION || newestVersion === "") {
+		logger.info("You are using the latest version");
```
Comment (kylecarbs) [inline packages/server/src/cli.ts]: Shouldn't output anything if latest imo

File: packages/server/src/updater.ts [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+			return 1;
+		}
+		if (aVal < bVal || isNaN(aVal)) {
+			return -1;
+		}
+	}
+};
+
+export const getRecentRelease = (): Promise<any> => {
+	return new Promise<string>((resolve, rej):void  => {
```
Comment (kylecarbs) [inline packages/server/src/updater.ts]: Should spell out `reject` fully since we spelt out `resolve` fully

File: packages/server/src/updater.ts [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+		const options = {
```
Comment (kylecarbs) [inline packages/server/src/updater.ts]: Should throw this directly into the `https.get`

File: packages/server/src/cli.ts [hunk 9]
```diff
@@ -139,6 +140,22 @@ if (isCli) {
 		}
 	}
 
+	let newestVersion = "";
+	if (process.env.VERSION) {
+		let currentVersion = process.env.VERSION;
+		const recentRelease = await getRecentRelease();
```
Comment (kylecarbs) [inline packages/server/src/cli.ts]: Needs to handle if this is rejected

File: packages/server/src/cli.ts [hunk 10]
```diff
@@ -139,8 +140,19 @@ if (isCli) {
 		}
 	}
 
-	logger.info(`\u001B[1mcode-server ${process.env.VERSION ? `v${process.env.VERSION}` : "development"}`);
+	if (isCli) {
+		let currentVersion = process.env.VERSION;
```
Comment (lsmoura) [inline packages/server/src/cli.ts]: This value is never reassigned. Use `const` instead?

```suggestion
		const currentVersion = process.env.VERSION;
```

File: packages/server/src/updater.ts [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+			if (res.statusCode !== 200) {
+				reject(Error("Failed to acquire release information"));
+			}
+			let body = "";
+			res.on("data", (chunk) => {
+				body += chunk;
+			});
+
+			res.on("end", () => {
+				let release = JSON.parse(body);
```
Comment (lsmoura) [inline packages/server/src/updater.ts]: This value is never reassigned. use `const` instead?

```suggestion
				const release = JSON.parse(body);
```

File: packages/server/src/cli.ts (hunk 9 above)
Comment (nol166) [reply packages/server/src/cli.ts]: made all these changes

**Comment by nol166** [comment]
resolved conflicts

**Pull request closed without merging**